layer/windows_writer.go

```diff
@@ -0,0 +1,95 @@
+package layer
+
+import (
+	"archive/tar"
+	"io"
+	"path"
+	"strings"
+)
+
+type WindowsWriter struct {
+	tarWriter          *tar.Writer
+	writtenParentPaths map[string]bool
+}
+
+func NewWindowsWriter(fileWriter io.Writer) *WindowsWriter {
+	return &WindowsWriter{
+		tarWriter:          tar.NewWriter(fileWriter),
+		writtenParentPaths: map[string]bool{},
+	}
+}
+
+func (w *WindowsWriter) Write(content []byte) (int, error) {
+	return w.tarWriter.Write(content)
+}
+
+func (w *WindowsWriter) WriteHeader(header *tar.Header) error {
```

Comment:
The behavior of `WriteHeader` when using the `WindowsWriter` is going to be different than that of the linux `tar.Writer` because this implementation writes **all** parent dir headers and `tar.Writer` does not. I think we either want a linux implementation that would, for example write `/cnb` if this was called with `/cnb/lifecycle` or we only want this implementation to only add the special dirs required by windows (I lean towards the latter so that users have control over the attributes of all normal files).

Reply:
I'll argue a bit that in the windows case all parent dirs are "special" because they're required for creating a valid layer tar. If we think of "writing everything necessary for a layer" as the feature that we want parity on between the different types of writers, then we can say that the two writers behave equivalently.

I would also lean toward the latter if I didn't see every parent dir as required/special, so my vote would be to leave as is. Thoughts?

Reply:
An alternative that we thought of was to create an interface that is more about writing layers, then create two concrete implementations, one for each OS. While I think that could be good to move towards in the long run, we opted not to do it right now to minimize the churn that would be required in dependent repos.